[IPSCCP] Variable not visible at Og.

CarlosAlbertoEnciso opened the pull request:
https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not create a constant expression for the initializer value.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

jryans commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
       SI->eraseFromParent();
     }
+
+    // Try to create a debug constant expression for the glbal variable
```
```suggestion
    // Try to create a debug constant expression for the global variable
```

CarlosAlbertoEnciso replied:
Fixed the typo.

jryans reviewed (approved):
This looks good to me overall, thanks for working on this improvement! 😄 

I noticed the PR title ends with a ":", perhaps some further words were lost or the ":" should be removed...?

I recommend waiting a few days for others to take a look as well.

nikic commented:
I don't think this chunk of code for converting a Constant into a DIExpression belongs into SCCP.cpp.

felipepiovezan commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+          InitIntOpt = API.tryZExtValue();
+        else if (auto Temp = API.trySExtValue(); Temp.has_value())
+          // Transform a signed optional to unsigned optional.
+          InitIntOpt = (uint64_t)Temp.value();
```
let's avoid C-style casts!

CarlosAlbertoEnciso replied:
Using instead `static_cast<uint64_t>`

felipepiovezan commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+        else if (auto Temp = API.trySExtValue(); Temp.has_value())
+          // Transform a signed optional to unsigned optional.
+          InitIntOpt = (uint64_t)Temp.value();
+        return DIB.createConstantValueExpression(InitIntOpt.value());
```
This is unconditionally calling `std::optional::value`, but the control flow above makes it look like we may have a `nullopt` here. This would be even more evident if we had not used an uninitialized variable here (it would be evident that we are missing an initialization case).

I think we need to account for the possibility of failure here:

```
DIExpression* createIntExpression(DIB, CV) {
        const APInt &API = dyn_cast<ConstantInt>(CV)->getValue();
        if (API.isNonNegative()) {
           if(optional<uint64_t> MaybeZExt = API.tryZExtValue())
                   return DIB.createConstantValueExpression(*MaybeZExt);
        }
        else if (optional<uint64_t> MaybeSExt = API.tryZExtValue())
                          return DIB.createConstantValueExpression(* MaybeSExt);
       return nullptr;
};

felipepiovezan replied:
Oh actually here we have another case where we should use `cast<ConstantInt>(CV)` instead of `dyn_cast`

CarlosAlbertoEnciso replied:
Changed to use `cast`.

felipepiovezan commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+      const Constant *CV = GV->getInitializer();
+      Type *Ty = GV->getValueType();
+      if (Ty->isIntegerTy()) {
+        GVEs[0]->replaceOperandWith(1, createIntExpression(CV));
```
Are we allowed to mutate a metadata like this? IIUC we can only do this if it is uniqued? Or not uniqued? I can never remember exactly

CarlosAlbertoEnciso replied:
The function `replaceOperandWith` takes that into consideration.

felipepiovezan commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+      if (Ty->isIntegerTy()) {
+        GVEs[0]->replaceOperandWith(1, createIntExpression(CV));
+      } else if (Ty->isFloatTy() || Ty->isDoubleTy()) {
+        const APFloat &APF = dyn_cast<ConstantFP>(CV)->getValueAPF();
```
a `dyn_cast` that is never checked should be a `cast`

CarlosAlbertoEnciso replied:
Changed to `cast`

felipepiovezan commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+          }
+        }
+      }
+    }
```
The control flow here is a bit difficult to follow, but I think we will improve this a lot once we make the helper function and replace some of those nested `if`s with early returns.

dwblaikie replied:
This could also be collapsed:
```
        } else {
          if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV)) {
            if (CE->getNumOperands() == 1) {
```
into this:
```
        } else if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV); CE->getNumOperands() == 1) {
```

But other refactoring might render that moot/be an alternative.

CarlosAlbertoEnciso replied:
Replaced with your suggested change.

CarlosAlbertoEnciso commented:
@jryans, @felipepiovezan, @nikic: Thanks for your comments.
I will address them and upload a new patch.

dwblaikie commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+          InitIntOpt = API.tryZExtValue();
+        else if (auto Temp = API.trySExtValue(); Temp.has_value())
+          // Transform a signed optional to unsigned optional.
+          InitIntOpt = (uint64_t)Temp.value();
```
Probably also favor `*` rather than `.value()` - not sure if we've clarified that in the style guide or not...

Oh, and the if condition could be simpler - no need to use `has_value` at all:
```
else if (auto Temp = API.trySExtValue())
```
Should suffice?

CarlosAlbertoEnciso commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
+    // initializer value.
+    SmallVector<DIGlobalVariableExpression *, 1> GVEs;
+    GV->getDebugInfo(GVEs);
+    if (GVEs.size() == 1) {
```
Good point.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Address reviewers comments:
- Refactor the new code into a helper function.
- Use early returns to reduce nested ifs.
- Remove C-style casts.
- Add validation to detect posible nullopt.

CarlosAlbertoEnciso commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -371,6 +372,52 @@ static bool runIPSCCP(
     }
+
+    // Try to create a debug constant expression for the glbal variable
+    // initializer value.
```
Moving the code to a helper function.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Address reviewers comments:
- Refactor the new code into a helper function.
- Use early returns to reduce nested ifs.
- Remove C-style casts.
- Add validation to detect posible nullopt.
- Collapse conditions in a single if.
- Remove 'has_value'.

CarlosAlbertoEnciso commented:
@jryans, @felipepiovezan, @nikic, @dwblaikie:
Thanks for all your feedback. I have uploaded a new patch that addresses all your points.
I am sorry if there are extra commits on this branch. This is the first update since moving to GitHub.

jryans reviewed (approved):
Looks good to me, thanks! 😄 

Other reviewers had more comments last time around, so probably best to wait for at least one more review before landing.

dwblaikie commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -106,14 +107,71 @@ static void findReturnsToZap(Function &F,
+      CE->getNumOperands() == 1) {
+    const Value *V = CE->getOperand(0);
+    const Constant *CV = dyn_cast<Constant>(V);
+    if (CV && !isa<GlobalValue>(CV);
```
I'm not sure this does what's intended (assuming the intent is that the `CV && !isa` part is meant to be part of the boolean condition of the `if` as well as the `CI` init) - these boolean expressions would have no effect, right? (the second part of the if, after the semicolon, would be the only condition?

felipepiovezan replied:
Yeah, I believe the intent here would be:

```
if (auto CI = dyn_cast_or_null<ConstantInt>(V)) {
  ...
}
```

Note that `!isa<GlobalValue>(CV)` is not necessary: a global value is always a pointer, therefore it cannot be a `ConstantInt`

CarlosAlbertoEnciso replied:
Good catch and nice simplification.

dwblaikie commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -106,14 +107,71 @@ static void findReturnsToZap(Function &F,
+    GVEs[0]->replaceOperandWith(1, DIB.createConstantValueExpression(0));
+    return;
+  }
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(CV);
```
This is probably problematic too - the initializer part of the `if` isn't part of the boolean condition - so the `CE` would be unconditionally dereferenced in the `CE->getNumOperands()` call - so either the `CE` init should use `cast`, not `dyn_cast` - or the condition needs to be `CE && CE->getNumOperands() ...` (or nested `if`s)

CarlosAlbertoEnciso replied:
Good catch. Replaced with nested `if`s.

dwblaikie replied:
This and the other conditions seemed pretty broken, glad they're now fixed - but is there some test coverage that was missed and would've demonstrated these problems/should be added to ensure comprehensive test coverage?

CarlosAlbertoEnciso replied:
I am happy to add some tests in [LocalTest.cpp](llvm/unittests/Transforms/Utils/LocalTest.cpp) to ensure we have test coverage for the new function.

nikic commented:
I still think that the logic for "convert Constant into DIExpression" should not be part of IPSCCP, but be a helper in some generic location (not sure where, debuginfo folks would know). This seems generically useful (to the point that I'm surprised it does not exist yet).

CarlosAlbertoEnciso commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -106,14 +107,71 @@ static void findReturnsToZap(Function &F,
+      InitIntOpt = API.tryZExtValue();
+    else if (auto Temp = API.trySExtValue())
+      // Transform a signed optional to unsigned optional.
+      InitIntOpt = static_cast<uint64_t>(*Temp);
```
Good point. I think the original code would be better if we are dealing with `APSInt`. But for this case, we have `APInt`.
Changed to 
```
    std::optional<int64_t> InitIntOpt = API.trySExtValue();
    return InitIntOpt ? DIB.createConstantValueExpression(
                            static_cast<uint64_t>(*InitIntOpt))
                      : nullptr;
```

jryans commented:
> I still think that the logic for "convert Constant into DIExpression" should not be part of IPSCCP, but be a helper in some generic location (not sure where, debuginfo folks would know). This seems generically useful (to the point that I'm surprised it does not exist yet).

I agree, it does seem good to make this available as a shared helper. Perhaps it could go to https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/Local.cpp? There are various shared debug info helpers stashed there already. (At some stage, a separate file just for debug info utils might be a good idea, but that can be separate task.)

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Address reviewers comments:
- Refactor the new code into a helper function.
- Use early returns to reduce nested ifs.
- Remove C-style casts.
- Add validation to detect posible nullopt.
- Collapse conditions in a single if.
- Remove 'has_value'.
- Additional if condition simplifications.
- Move new logic into a general helper function.

CarlosAlbertoEnciso commented:
Uploaded new patch with the new logic moved to https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/Local.cpp

felipepiovezan commented on llvm/lib/Transforms/IPO/SCCP.cpp:
```diff
@@ -106,14 +107,14 @@ static void findReturnsToZap(Function &F,
-    std::function<DominatorTree &(Function &)> GetDT,
-    std::function<BlockFrequencyInfo &(Function &)> GetBFI,
-    bool IsFuncSpecEnabled) {
+static bool
```
this whitespace diff should not be part of the patch

CarlosAlbertoEnciso replied:
Removed.

jryans commented on llvm/include/llvm/Transforms/Utils/Local.h:
```diff
@@ -515,6 +515,14 @@ Value *invertCondition(Value *Condition);
 bool inferAttributesFromOthers(Function &F);
 
+///===---------------------------------------------------------------------===//
+///  DebugInfo utilities
```
I would think it fits in with the existing "Dbg Intrinsic utilities" section:

https://github.com/llvm/llvm-project/blob/df8e0d057d67a8167f2bdc81d8b61033beb55da5/llvm/include/llvm/Transforms/Utils/Local.h#L258

Perhaps it's better to append it to the end of that section instead of creating a new, similar section?

felipepiovezan commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3606,3 +3606,41 @@ bool llvm::inferAttributesFromOthers(Function &F) {
   return Changed;
 }
+
+DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant *C,
```
We never check if `C` or `Ty` are null, we can make these references

CarlosAlbertoEnciso replied:
Passing them as references.

felipepiovezan reviewed (approved):
I left two other minor comments, otherwise the patch LGTM!

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Address reviewers comments:
- Refactor the new code into a helper function.
- Use early returns to reduce nested ifs.
- Remove C-style casts.
- Add validation to detect posible nullopt.
- Collapse conditions in a single if.
- Remove 'has_value'.
- Additional if condition simplifications.
- Move new logic into a general helper function.
- Move new logic to the 'Dbg Intrinsic utilities'" section.
- Pass 'Constant' and 'Type' as references.
- Remove extra whitespace differences (incorrect clang-format).

CarlosAlbertoEnciso commented:
@jryans, @felipepiovezan: uploaded new patch to reflect your latest comments.
@nikic, @dwblaikie: do you have any additional comments?

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3074,6 +3074,44 @@ void llvm::hoistAllInstructionsInto(BasicBlock *DomBlock, Instruction *InsertPt,
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(&C)) {
+    if (CE->getNumOperands() == 1) {
+      const Value *V = CE->getOperand(0);
+      if (auto CI = dyn_cast_or_null<ConstantInt>(V)) {
```
Omit braces

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3074,6 +3074,44 @@ void llvm::hoistAllInstructionsInto(BasicBlock *DomBlock, Instruction *InsertPt,
+    return DIB.createConstantValueExpression(0);
+
+  if (const ConstantExpr *CE = dyn_cast<ConstantExpr>(&C)) {
+    if (CE->getNumOperands() == 1) {
```
```suggestion
    if (CE->getOpcode() == Instruction::IntToPtr) {
```
I *think* this is what you're actually trying to handle here? I think this would be clearer than identifying inttoptr as "an expression with one operand that has pointer result and ConstantInt operand".

nikic commented on llvm/test/Transforms/SCCP/pr50901.ll:
```diff
@@ -0,0 +1,184 @@
+  %6 = load ptr, ptr @_ZL4g_77, align 8, !dbg !74, !tbaa !71
+  store ptr %6, ptr @g_7, align 8, !dbg !75, !tbaa !71
+  ret void, !dbg !76
+}
```
You should be able to drop all the `!tbaa` metadata (really, any non-dbg metadata) from this test. It shouldn't be relevant for IPSCCP.

CarlosAlbertoEnciso replied:
Removed all the `!tbaa` metadata.

nikic commented on llvm/test/Transforms/SCCP/pr50901.ll:
```diff
@@ -0,0 +1,184 @@
+; CHECK-DAG: ![[DBG7]] = distinct !DIGlobalVariable(name: "g_77", {{.*}}
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-linux-gnu"
```
Data layout and triple shouldn't be necessary. (If you specify a triple, it's either useless or you need to add REQUIRES, otherwise you will break the build.)

CarlosAlbertoEnciso replied:
Removed `datalayout` and `triple`.

nikic reviewed (commented):
Looks fine, just some nits.

djtodoro commented on llvm/test/Transforms/SCCP/pr50901.ll:
```diff
@@ -0,0 +1,184 @@
+
+define dso_local void @_Z3barv() !dbg !46 {
+entry:
+  %0 = load i32, ptr @_ZL4g_11, align 4, !dbg !49, !tbaa !50
```
I believe that you could use only one `!dbg` for all the instructions in the function and get rid of all the others.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Address reviewers comments:
- Refactor the new code into a helper function.
- Use early returns to reduce nested ifs.
- Remove C-style casts.
- Add validation to detect posible nullopt.
- Collapse conditions in a single if.
- Remove 'has_value'.
- Additional if condition simplifications.
- Move new logic into a general helper function.
- Move new logic to the 'Dbg Intrinsic utilities'" section.
- Pass 'Constant' and 'Type' as references.
- Remove extra whitespace differences (incorrect clang-format).
- Use 'Instruction::IntToPtr' with pointer types.
- Reduce lit test case.
- Add unittests for the new helper function.

CarlosAlbertoEnciso commented:
Thanks for all your comments.
Updated to address:
@dwblaikie: include unittests (Covering the most common types for constants).
@djtodoro and @nikic: reduce the lit test (Removing redundant `!dbg`).
@nikic: extra braces and different `if` condition for pointer types.

dwblaikie commented:
(nothing further from me, but I'll leave it to other folks more familiar with this to approve)

CarlosAlbertoEnciso commented:
> (nothing further from me, but I'll leave it to other folks more familiar with this to approve)

Thanks

CarlosAlbertoEnciso commented:
Thanks to all reviewers for your valuable comments.
Any additional points to address, before I merge the changes.

nikic commented:
I think you can go ahead and merge this. My comments have been addressed, and other people already approved this previously.

Endilll commented:
Looks like this breaks OpenCL: https://github.com/llvm/llvm-project/issues/50901#issuecomment-1870283804
